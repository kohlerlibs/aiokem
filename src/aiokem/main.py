'''Main File'''

from aiohttp import ClientTimeout, ClientSession

AUTHENTICATION_URL = 'https://kohler-homeenergy.okta.com/oauth2/default/v1/token'
CLIENT_KEY = "MG9hMXFpY3BkYWdLaXdFekYxZDg6d3Raa1FwNlY1T09vMW9PcjhlSFJHTnFBWEY3azZJaXhtWGhINHZjcnU2TWwxSnRLUE5obXdsMEN1MGlnQkVIRg=="
API_KEY = "pgH7QzFHJx4w46fI~5Uzi4RvtTwlEXp"
API_BASE = "https://az-amer-prod-hems-capp.jollyglacier-72224ec0.eastus.azurecontainerapps.io"

class AioKem:
    def __init__(self, session: ClientSession = None) -> None:
        '''Initialize the AioKem class'''
        self._token: str|None = None
        self._refresh_token: str|None = None
        self._session = session if session else ClientSession(
            timeout=ClientTimeout(total=10)
        )

    async def login(self, username: str, password: str) -> None:
        '''Login to the server'''
        headers = {
            'accept': 'application/json',
            'authorization': 'Basic MG9hMXFpY3BkYWdLaXdFekYxZDg6d3Raa1FwNlY1T09vMW9PcjhlSFJHTnFBWEY3azZJaXhtWGhINHZjcnU2TWwxSnRLUE5obXdsMEN1MGlnQkVIRg==',
            'content-type': 'application/x-www-form-urlencoded',
        }
        data = {
            'grant_type': 'password',
            'username': username,
            'password': password,  # Replace 'pass' with the actual password
            'scope': 'openid profile offline_access email'
        }

        response = await self._session.post(AUTHENTICATION_URL, headers=headers, data=data)
        response_data = await response.json()
        if response.status != 200:
            raise Exception(f"Login failed: {response_data.get('error_description', 'Unknown error')}")
        self._token = response_data.get('access_token')
        self._refresh_token = response_data.get('refresh_token')
        if not self._token:
            raise Exception("Login failed: No access token received")
        if not self._refresh_token:
            raise Exception("Login failed: No refresh token received")


    async def get_homes(self) -> dict:
        url = f'{API_BASE}/kem/api/v3/homeowner/homes'
        headers = {
            'apikey': API_KEY,
            'authorization': f'bearer {self._token}',
        }

        response = await self._session.get(url, headers=headers)
        if response.status != 200:
            raise Exception(f"Failed to fetch homes: {response.status}")
        response_data = await response.json()
        if not response_data:
            raise Exception("No data received")
        return response_data

    async def get_generator_data(self, generator_id: int) -> dict:
        url = f'{API_BASE}/kem/api/v3/devices/{generator_id}'
        headers = {
            'apikey': API_KEY,
            'authorization': f'bearer {self._token}',
        }

        response = await self._session.get(url, headers=headers)
        if response.status != 200:
            raise Exception(f"Failed to fetch generator data: {response.status}")
        response_data = await response.json()
        if not response_data:
            raise Exception("No data received")
        return response_data
    
    async def close(self) -> None:
        '''Close the session'''
        await self._session.close()

