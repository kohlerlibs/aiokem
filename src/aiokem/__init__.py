__version__ = "1.0.0-rc.1"

from .exceptions import (
    AioKemError,
    AuthenticationCredentialsError,
    AuthenticationError,
    CommunicationError,
    ServerError,
)
from .main import AioKem

__all__ = (
    "AioKem",
    "AioKemError",
    "AuthenticationCredentialsError",
    "AuthenticationError",
    "CommunicationError",
    "ServerError",
)
