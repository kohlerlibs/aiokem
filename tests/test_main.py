from http import HTTPStatus
from unittest.mock import AsyncMock, Mock

import pytest
from aiohttp import ClientConnectionError

from aiokem.exceptions import (
    AuthenticationCredentialsError,
    AuthenticationError,
    CommunicationError,
)
from aiokem.main import AUTHENTICATION_URL, AioKem


@pytest.mark.asyncio
async def test_login():
    # Create a mock session
    mock_session = Mock()
    mock_session.post = AsyncMock()
    kem = AioKem(session=mock_session)

    # Mock the response for the login method
    mock_response = AsyncMock()
    mock_response.status = 200
    mock_response.json.return_value = {
        "access_token": "mock_access_token",
        "refresh_token": "mock_refresh_token",
    }
    mock_session.post.return_value = mock_response

    # Call the login method
    await kem.login("username", "password")

    # Assert that the access token and refresh token are set correctly
    assert kem._token == "mock_access_token"  # noqa: S105
    assert kem._refresh_token == "mock_refresh_token"  # noqa: S105
    # Assert that the session.post method was called with the correct URL and data
    mock_session.post.assert_called_once()
    assert mock_session.post.call_args[0][0] == AUTHENTICATION_URL
    assert mock_session.post.call_args[1]["data"] == {
        "grant_type": "password",
        "username": "username",
        "password": "password",
        "scope": "openid profile offline_access email",
    }


@pytest.mark.asyncio
async def test_login_exceptions():
    # Create a mock session
    mock_session = Mock()
    mock_session.post = AsyncMock()
    kem = AioKem(session=mock_session)

    # Mock the response for the login method
    mock_response = AsyncMock()
    mock_response.status = HTTPStatus.BAD_REQUEST
    mock_response.json.return_value = {
        "error_description": "The credentials provided were invalid.",
    }
    mock_session.post.return_value = mock_response

    # Call the login method
    with pytest.raises(AuthenticationCredentialsError) as excinfo:
        await kem.login("username", "password")

    # Assert that the access token and refresh token are set correctly
    assert kem._token is None
    assert kem._refresh_token is None
    # Assert that the exception message is correct
    assert (
        str(excinfo.value)
        == "Invalid Credentials: The credentials provided were invalid. Code 400"
    )

    mock_response = AsyncMock()
    mock_response.status = HTTPStatus.FORBIDDEN
    mock_response.json.return_value = {
        "error_description": "Disallowed operation.",
    }
    mock_session.post.return_value = mock_response
    # Call the login method
    with pytest.raises(AuthenticationError) as excinfo:
        await kem.login("username", "password")
    assert str(excinfo.value) == "Authentication failed: Disallowed operation. Code 403"

    mock_session.post.side_effect = ClientConnectionError("Internet connection error")

    # Call the login method
    with pytest.raises(CommunicationError) as excinfo:
        await kem.login("username", "password")
    assert str(excinfo.value) == "Connection error: Internet connection error"
