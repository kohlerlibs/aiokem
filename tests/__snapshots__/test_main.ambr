# serializer version: 1
# name: test_get_generator_data
  dict({
    'addedBy': None,
    'associatedUsers': list([
      'pete.rage@rage.com',
    ]),
    'batteryVoltageV': 13.9,
    'connectionType': 'Unknown',
    'controllerClockTimestamp': '2025-04-15T07:08:50',
    'controllerTempF': 71.6,
    'coolingType': 'Air',
    'device': dict({
      'address': dict({
        'address1': 'Highway 66',
        'address2': None,
        'city': 'Somewhere',
        'country': 'US',
        'lat': 41.3341111,
        'long': -72.3333111,
        'postalCode': '00000',
        'state': 'CA',
      }),
      'alertCount': 0,
      'controllerType': 'RDC2 (Blue Board)',
      'currentFirmware': 'RDC2.4 3.4.5',
      'dealerOrgs': list([
        dict({
          'address': dict({
            'address1': 'Highway 66',
            'address2': None,
            'city': 'Revisited',
            'country': None,
            'lat': None,
            'long': None,
            'postalCode': '000000',
            'state': 'CA',
          }),
          'adminEmails': list([
            'admin@gmail.com',
          ]),
          'businessPartnerNo': '123456',
          'deviceCount': 71,
          'displayPhoneNumber': '(999) 999-9999',
          'e164PhoneNumber': '+199999999999',
          'id': 123,
          'name': 'Generators R Us',
          'technicianCount': 3,
          'userCount': 4,
          'wizardComplete': True,
          'wizardStep': 'OnboardingComplete',
        }),
      ]),
      'deviceHost': 'Oncue',
      'deviceIpAddress': '1.1.1.1:2402',
      'displayName': 'Generator 1',
      'firmwareVersion': '3.4.5',
      'hasAcceptedPrivacyPolicy': True,
      'hasServiceAgreement': None,
      'id': 12345,
      'isConnected': True,
      'lastConnectedTimestamp': '2025-04-14T09:30:17+00:00',
      'lastMaintenanceTimestamp': '2025-04-10T09:12:59-04:00',
      'macAddress': '00:10:63:20:E1:91',
      'maintenancePeriodDays': 365,
      'model': 'Model20KW',
      'modelDisplayName': '20 KW',
      'nextMaintenanceTimestamp': '2026-04-10T09:12:59-04:00',
      'product': 'Rdc2v4',
      'productDisplayName': 'RDC 2.4',
      'serialNumber': '123MGVHR4567',
      'status': 'ReadyToRun',
      'statusUpdateTimestamp': '2025-04-14T09:29:01+00:00',
      'totalRuntimeHours': 120.2,
    }),
    'engineCompartmentTempF': None,
    'engineCoolantTempF': None,
    'engineFrequencyHz': 0,
    'engineOilPressureOk': True,
    'engineOilPressurePsi': None,
    'engineSpeedRpm': 0,
    'engineState': 'Standby',
    'engineStateDisplayNameEn': 'Standby',
    'exercise': dict({
      'durationMinutes': 20,
      'frequency': 'Weekly',
      'lastEndTimestamp': '2025-04-12T14:19:59+00:00',
      'lastStartTimestamp': '2025-04-12T14:00:00+00:00',
      'mode': 'Unloaded',
      'nextStartTimestamp': '2025-04-19T10:00:00-04:00',
      'runningMode': None,
    }),
    'fuelType': 'LiquidPropane',
    'generatorLoadPercent': 0,
    'generatorLoadW': 0,
    'generatorVoltageAvgV': 0,
    'lastRanTimestamp': '2025-04-12T14:00:00+00:00',
    'loadShed': dict({
      'isConnected': True,
      'parameters': list([
        dict({
          'definitionId': 1,
          'displayName': 'HVAC A',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 2,
          'displayName': 'HVAC B',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 3,
          'displayName': 'Load A',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 4,
          'displayName': 'Load B',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 5,
          'displayName': 'Load C',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 6,
          'displayName': 'Load D',
          'isReadOnly': False,
          'value': False,
        }),
      ]),
    }),
    'lubeOilTempF': 42.8,
    'pim': dict({
      'isConnected': False,
      'parameters': list([
        dict({
          'definitionId': 7,
          'displayName': 'Digital Output B1 Value',
          'isReadOnly': True,
          'value': False,
        }),
        dict({
          'definitionId': 8,
          'displayName': 'Digital Output B2 Value',
          'isReadOnly': True,
          'value': False,
        }),
        dict({
          'definitionId': 9,
          'displayName': 'Digital Output B3 Value',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 10,
          'displayName': 'Digital Output B4 Value',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 11,
          'displayName': 'Digital Output B5 Value',
          'isReadOnly': False,
          'value': False,
        }),
        dict({
          'definitionId': 12,
          'displayName': 'Digital Output B6 Value',
          'isReadOnly': False,
          'value': False,
        }),
      ]),
    }),
    'powerSource': 'Utility',
    'remoteResetCounterSeconds': 0,
    'runtimeSinceLastMaintenanceHours': 0.3,
    'serverIpAddress': '2.2.2.2',
    'serviceAgreement': dict({
      'beginTimestamp': None,
      'hasServiceAgreement': None,
      'term': None,
      'termDays': None,
      'termMonths': None,
    }),
    'setOutputVoltageV': 240,
    'switchState': 'Auto',
    'totalOperationHours': 33932.3,
    'totalRuntimeHours': 120.2,
    'utilityVoltageV': 259.7,
  })
# ---
# name: test_get_homes
  list([
    dict({
      'address': dict({
        'address1': 'Highway 66',
        'address2': None,
        'city': 'Somewhere',
        'country': 'US',
        'lat': 41.334111,
        'long': -72.3333111,
        'postalCode': '000000',
        'state': 'CA',
      }),
      'devices': list([
        dict({
          'address': dict({
            'address1': 'Highway 66',
            'address2': None,
            'city': 'Somewhere',
            'country': 'US',
            'lat': 41.334111,
            'long': -72.3333111,
            'postalCode': '000000',
            'state': 'CA',
          }),
          'alertCount': 0,
          'controllerType': 'RDC2 (Blue Board)',
          'currentFirmware': 'RDC2.4 3.4.5',
          'dealerOrgs': list([
            dict({
              'address': dict({
                'address1': 'Highway 66',
                'address2': None,
                'city': 'Revisited',
                'country': None,
                'lat': None,
                'long': None,
                'postalCode': '000000',
                'state': 'CA',
              }),
              'adminEmails': list([
                'admin@gmail.com',
              ]),
              'businessPartnerNo': '123456',
              'deviceCount': 71,
              'displayPhoneNumber': '(999) 999-9999',
              'e164PhoneNumber': '+199999999999',
              'id': 123,
              'name': 'Generators R Us',
              'technicianCount': 3,
              'userCount': 4,
              'wizardComplete': True,
              'wizardStep': 'OnboardingComplete',
            }),
          ]),
          'deviceHost': 'Oncue',
          'deviceIpAddress': '1.1.1.1:2402',
          'displayName': 'Generator 1',
          'firmwareVersion': '3.4.5',
          'hasAcceptedPrivacyPolicy': True,
          'hasServiceAgreement': None,
          'id': 12345,
          'isConnected': True,
          'lastConnectedTimestamp': '2025-04-14T09:30:17+00:00',
          'lastMaintenanceTimestamp': '2025-04-10T09:12:59',
          'macAddress': '00:10:63:20:E1:91',
          'maintenancePeriodDays': 365,
          'model': 'Model20KW',
          'modelDisplayName': '20 KW',
          'nextMaintenanceTimestamp': '2026-04-10T09:12:59',
          'product': 'Rdc2v4',
          'productDisplayName': 'RDC 2.4',
          'serialNumber': '123MGVHR4567',
          'status': 'ReadyToRun',
          'statusUpdateTimestamp': '2025-04-14T09:29:01+00:00',
          'totalRuntimeHours': 120.2,
        }),
      ]),
      'id': 12345,
      'name': 'Generator 1',
      'weatherCondition': 'Mist',
      'weatherTempF': 46.11200000000006,
      'weatherTimePeriod': 'Day',
    }),
  ])
# ---
